### ISSUE TITLE
Add support for importing USD scenes

### ISSUE THREAD

**Issue by ywmaa** [godotengine/godot-proposals#7744]
### Describe the project you are working on

a 3D shooter game

### Describe the problem or limitation you are having in your project

~~my level is made using geometry nodes in blender, blender got the option to export the level in USD, and instead of duplicating the instances of the same mesh, it includes the instances as references to the Original mesh in the USD format.~~
Edit: This previous issue is fixed by blender, see this issue:
https://github.com/godotengine/godot-proposals/issues/8991

This issue is now more about whether Godot should support USD by default because it is becoming industry standard or not ?

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

support the USD format. ~~with care support for importing "instances" as instances instead of duplicating.~~

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

N/A

### If this enhancement will not be used often, can it be worked around with a few lines of script?

can't be done with few lines of script.

### Is there a reason why this should be core and not an add-on in the asset library?

it can be done as an addon, but having it built-in like GLTF/GLB will make it maintained often, and it may become the second most used format in Godot Engine after the GLTF format.

also I think it is becoming an industry standard like FBX ?
so it will make Godot Engine work easy with almost all 3D software.

Edit:
Re considered the actual purpose of this Issue.

**Comment by Calinou** [comment]
There are 2 approaches for this:

- Use Blender to import USD, like we do for .blend files. This wouldn't allow for run-time loading in exported projects though.
- Use [tinyusdz](https://github.com/syoyo/tinyusdz), which is a standalone library.

**Comment by Calinou** [comment]
> And AFAIK GLTF doesn't handle instancing/instances.

I'm pretty sure there is a glTF extension for that already :slightly_smiling_face:

**Comment by ywmaa** [comment]
> I'm pretty sure there is a glTF extension for that already

Well if that's true, then I don't mind importing USD by converting from USD to GLTF using blender.

Though I couldn't see an option for exporting instances in the GLTF exporter in blender.

Maybe they didn't implement it.

**Comment by ywmaa** [comment]
> Maybe they didn't implement it.

turns out I am wrong, blender autmatically packages all instances as somewhat references to one mesh.

see this screenshot from a GLTF File that I exported from blender :
![image](https://github.com/godotengine/godot-proposals/assets/68971808/75c100af-adf1-4480-8dec-dc47ad208dbf)

the problem is that Godot Doesn't import them as instances, instead Godot imports them as standalone mesh data : 
![image](https://github.com/godotengine/godot-proposals/assets/68971808/063ef1ce-b068-4c4e-9b0c-518bfcf783e9)


but I think this should work using https://github.com/godotengine/godot/issues/56312
and not this issue.

## Also About this issue

> glTF materials are supported by Godot's importer.

It imports the material if it is using simple PBR setup, or using textures.
but a complex PBR setup using the principled bsdf in blender, won't be imported automatically, it will need to be redone.

**Comment by NicTanghe** [comment]
Run-time loading in USD files from the web is a game changer.
I highly advise some hero to not take shortcuts and support the full thing.
Would be a bit weird if your browser didn’t fully support HTML.

**Comment by ywmaa** [comment]
> I highly advise some hero to not take shortcuts and support the full thing.
lol I was working on it using the blender shortcut, but anyway this problem is not only about getting the code, also about the maintainers and the community overall accepts increasing Godot Engine size for USD importing support.

**Comment by Kimau** [comment]
Listening to the talk from Godotcon it does sound like USD and MaterialX would have solved a lot of the instancing issues and blender porting shaders issues.
Also the .usd variant flow is super powerful for game editors.
All of which TinyUSD seems to support, though I've not used it.

I know that most the industry is going all in on USD at least the 3d/vfx folks. Adobe/Substance has made strong commitment also other engines (Unity, Unreal) and DCI (Maya, Houdini, Blender ect...). It's defo coming up like the standard scene format like gltf seems ot be replacing fbx as standard model format.

**Comment by ywmaa** [comment]
I am not sure of the state of this proposal, it seems more people are with the idea of having a native usd importer like tinyusd.

Do you think it is worth that I try doing it and open a pull request ?
Or wait a but to see if this issue gains more attention ?

**Comment by PierceLBrooks** [comment]
@ywmaa I am a fan of the existing "`GLTFDocument`" API provided by Godot, and would definitely really appreciate seeing a comparable parallel USD/USDZ "Document" class become available for making apps that have a need for runtime asset loading. The expression of animation skeletons/skins in GLTF is a bit weak, and USD's format seems stronger in that domain so I'd probably opt for it more often personally.

**Comment by digisomni** [comment]
We love glTF/GLB, but realize that USD is better equipped to handle the idea of a "full scene" as it attempts to carry over even shaders. This is extremely important for any real production pipeline where there are editors and engines involved. Artists might opt to create in Unity and deploy in Godot, and be perfectly happy in doing so.

It also allows us to utilize tooling without vendor lock-in. Because the practical alternative to USD support is to enforce that the editor is the same as the deployment environment, thus building out tooling exclusively for one platform e.g. Godot, Unreal, Unity, etc.

**Comment by InfernalWAVE** [comment]
I think this is a good idea. Apple seems to be going forward with the USDZ file format, which is just compressed USD. Apple tends to influence the industry. For example, in order to texture paint in 4K in Procreate, you need to import/export using USDZ format.

**Comment by blikpils** [comment]
I am kind of surprised that this proposal has not generated more backing. USD and MaterialX are the next industry standards and having this support will be crucial. Just being able to export an entire level, load only a portion you need in Houdini, add procedural content or vfx and export it back to Godot would be gamechanger.

**Comment by InfernalWAVE** [comment]
Juan does not seem to be a fan: https://x.com/reduzio/status/1770049465866432656

**Comment by shevisj** [comment]
Adding some noise here as I think there's a pretty big opportunity cost to not supporting USD natively. There is obviously a lot of momentum in the industry to standardize around it which opens up interoperability, but crucially there is also a lot of USD-specific tooling being created which we could really benefit from (or miss out on). While extension support could potentially plug that gap, I think that native support would position Godot as a more serious contender in larger studios due to the interop and tooling it unlocks.

**Comment by SoulSloth** [comment]
Definitely would love to see Godot implement OpenUSD. It'd be a lot of fun writing tools to do generative meshes and have them be engine agnostic with one export.

**Comment by hqantico** [comment]
I would love to see USD support for Godot. It's becoming more and more adopted across the industry and really opens up unique workflows with assets and layout.

**Comment by hqantico** [comment]
I would love to see USD support for Godot. It's becoming more and more adopted across the industry and really opens up unique workflows with assets and layout.

**Comment by Calinou** [comment]
@chipper-compatriot Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by Aiacos** [comment]
I've opened a PR implementing native USD import support for Godot: https://github.com/godotengine/godot/pull/115756

The implementation uses [tinyusdz](https://github.com/syoyo/tinyusdz) (Apache 2.0, dependency-free C++14) as the USD parser, following the same module architecture as the existing glTF and FBX importers.

**What's currently supported:**
- Import of `.usd`, `.usda`, `.usdc`, `.usdz` files via `EditorSceneFormatImporter`
- Meshes with vertices, normals, UVs, and N-gon fan triangulation
- UsdPreviewSurface → StandardMaterial3D material conversion
- MaterialX XML parsing and compilation to Godot spatial shaders
- Lights (distant, sphere, disk, rect, dome) and cameras
- Automatic coordinate system conversion (Z-up cm → Y-up m)

**Planned for follow-up work:**
- Skeletal animation (UsdSkel) — stub classes are in place
- Transform/property animation
- Point instancing
- Layer composition (sublayers, references)
- USD export

Feedback and testing with real-world USD assets would be greatly appreciated!

**Comment by AndreBorrmann** [comment]
The following Godot plugin [IDTX-Flow](https://github.com/Immersive-Data-Center-Management/idtx-flow) has been released recently to enable the usage of openUSD files within Godot at editor and gameplay runtime. In its current iteration it is capable to import USD stages into Godot nodes.

You would add a `UsdStageNode3D` node to your Godot scene and set the USD stage file it shall import in the node inspector. Choose from your `res://` project folder or a remote (http / https) resource. The plugin is under active development and will further grow. Even though the feature list is not yet finally planned out, you might want to keep an eye or even submit ideas or contribute.

**Comment by fire** [comment]
I have tested [IDTX-Flow](https://github.com/Immersive-Data-Center-Management/idtx-flow) and captures the intent of adding openusd to Godot Engine.

As with any new project it has bugs, but the base concept is implemented.

**Comment by fire** [comment]
I propose to close this issue.

**Comment by TheDarkThief** [comment]
I'd like this to be a native feature rather than an extension

**Comment by fire** [comment]
I am supportive but there is push-back for adding like 3.63+ megabytes to the Godot Engine editor and templates.

https://github.com/Immersive-Data-Center-Management/idtx-flow/releases/tag/v0.4.0

<img width="1056" height="221" alt="Image" src="https://github.com/user-attachments/assets/aee475d5-d797-427b-9c89-d172d5e87f78" />

@Calinou Can you give me a second opinion?

**Comment by Calinou** [comment]
> @Calinou Can you give me a second opinion?

If this is editor-only, I'd say this is acceptable, but it's definitely too much for export templates (unless opt-in for custom builds, similar to tinyexr before 4.6).

I'm not sure how many people would load (or save) USD scenes at runtime regardless. glTF seems more popular for UGC interchange right now.

**Comment by fire** [comment]
@AndreBorrmann I am interested in an editor only contribution to Godot Engine, but as far as I know that goes against the design of OpenUSD can you comment?

Edited:

Like without a runtime openusd node, we would be required to bake openusd scenes to godot scene format.

**Comment by AndreBorrmann** [comment]
Hi there,

I'm not in any position to provide any specifics whether the "design of OpenUSD" is suited better for design time only or design time + runtime in game development. However, imho the idea of openUSD is to provide a standard format that is platform independent and decoupled from its usage in any tool or application. It has been born at Pixar, as far as I'm aware, to support their film production pipeline that till then used many tool with many different data and asset formats. Having them unified was a big step into streamlining and optimizing their production pipelines.

Having said this, I'd love to see Godot natively supporting openUSD. However, "supporting" openUSD comes with some flavors to it. Thinking of game development, it might be just fine to treat usd files as simple asset format that gets imported once into the engine and that's it. Everything else is done in the game engine itself. There might be "delta-imports" required, in case the asset got optimized, but this could be it. Having only this feature would be a good argument to provide this as Godot editor feature only.

Games often use DLC's or provide features, addOns, new levels or whatever to keep players engaged or be able to sell more content. USD could be used here to provide a platform agnostic format that ships to whatever target hardware. Latest here, some sort of runtime support would be required.

In addition to this, our current (the project/organization I'm working for) aim is to allow the creation of enterprise applications using Godot. The biggest difference between a game and an enterprise application, from my point of view, is, that in the latter one, the content is mainly volatile and data driven and can't be baked into or shipped with the final game/binary. This means, while designing the application/experience there might never be the full set of assets available that are required to be rendered at runtime. Thus the requirement to be able to add content/assets at runtime is essential in this kind of use case. Further more enterprise application adoption rises if it relies on industry standards that vendors of assets can easily adhere to. Nevertheless, there are many more advantages that openUSD as framework provides in the context of enterprise applications I'm not bothering you with...

Let's assume there is no further question, that openUSD shall be supported natively within Godot for now. What are the implications to the engine?

The biggest concern raised is the size bloat. Well, the compiled openUSD library required to support openUSD features would be the biggest part and vary between 50MB and up to 600MB on Windows debug builds. This feels heavy at first sight, but assuming games that are content rich with fine grained 3D models and high fidelity textures, they will easily sum up to some GB in size if not carefully optimized. I'd argue, that the library size and the increase of the runtime binary in this scenario are neglect-able.

TL;DR
From my point of view, adding openUSD natively to Godot for design time (editor) and runtime should not be discussed purely from the size bloat point of view. Adding openUSD to Godot natively should be considered high priority on the roadmap of the engine.

Looking left/right, Epic recently announced their way to UE 6 and that they will move the engine into the direction of using open standards and openUSD is one of them, that will become first-class citizen:
```
We’re going to do this with a determined effort to move UE beyond just extensibility and into open specifications for interoperability.
Where existing standards such as glTF or USD are capable of fulfilling UE6’s needs, we will make them first-class formats within the
engine.
```
Source: https://www.unrealengine.com/news/the-road-to-ue-6

And Remedy Games also proved using openUSD in game engines is a viable decision:
https://gdcvault.com/play/1034201/Building-Game-Worlds-with

**Comment by dsnopek** [comment]
Raw size isn't the only issue here. There's also the extra maintenance burden that this puts on Godot maintainers, weighed against how many Godot users will take advantage of the USD format.

I personally think USD support should be provided as a GDExtension (or addon), at least  for now, and there's already a couple of options:

- https://github.com/tefusion/godot-usd - a GDExtension using tinyusdz
- https://github.com/V-Sekai/godot-usd - an addon that uses Blender to automatically convert to GLTF on import

If we reach a point where USD is as popular as GLTF and FBX, that could be an argument merging for into Godot itself.

However, Godot does need to draw some lines, and sometimes allowing enthusiastic sub-communities to support a feature can lead to better solutions that are more responsive to user needs, than putting something into Godot itself (where things move slower and maintainers have many other considerations).

**Comment by NicTanghe** [comment]
OK but logically speaking all users should be taking advantage of USD and ditch all the other redundant/fragmented/inferior/slower standards.

**Comment by AThousandShips** [comment]
Godot should offer support for what people are using, we cannot and should not preempt this or try to somehow direct the use of file formats, especially if the formats are not widely used and that leads to less active testing of the format. As a general rule features that are not widely demanded are risky not just because they require resources to be unevenly used, but because features that most don't use don't get tested enough

Meanwhile extensions are a perfect testing ground that allows the feature to be tested and grow, with a more dedicated userbase and maintainers, and when the format is mature and widely spread enough it can be integrated

**Comment by fire** [comment]
@dsnopek I've been trialing foundational openusd support via @AndreBorrmann 's gdextension work at https://github.com/Immersive-Data-Center-Management/idtx-flow

idtx-flow has some flaws, but they're not fatal. They can worked on via idtx-flow issues and pull requests.

Edited:

The evaluation for tinyusd based code systems is that it doesn't import correctly, and my code requires a Blender download which is much bigger than an openusd sdk.

**Comment by NicTanghe** [comment]
> Godot should offer support for what people are using, we cannot and should not preempt this or try to somehow direct the use of file formats, especially if the formats are not widely used and that leads to less active testing of the format. As a general rule features that are not widely demanded are risky not just because they require resources to be unevenly used, but because features that most don't use don't get tested enough
> 

people can't use things that aren't supported you cant only start selling people cars after they have decided to drive cars instead of horses they need roads and cars to be available first.

Im not saying ditch all the fragmented standards im saying please do the thing propperly.

**Comment by Aiacos** [comment]
I would like to add that might people not using usd because USD it is not
just a conversion tool or file format like FBX or alembic, it is a full
featured framework to interchange data (assets, shading, lighting,
settings, shots) layer based. That is what a USD user want to see, a proper
implementation like in Houdini or Maya

Thanks,
L

Il giorno ven 26 giu 2026 alle 20:32 Hameron ***@***.***> ha
scritto:

**Comment by AndreBorrmann** [comment]
Hi there,

even though I'm totally fine in keeping USD import as an extension I'd still argue, that the roadmap should discuss a proper and deep integration and like @Aiacos mentioned, with a deep integration into the engine it can facilitate the features openUSD provides. Providing a proper user experience around this could be achieved with an add-on for sure, but a direct integration would benefit both, the implementation and the user experience.

Without repeating myself, I still would argue, that making a bold step here could future proof Godot, to allow building enterprise applications and be ready for a whole new marked.

**Comment by AThousandShips** [comment]
Godot development should follow where the community leads and adapt to the *current* needs of the community, not lead the way with potential future needs or try to guide the tools or resources the community *should* use, if this becomes popular as an add-on, and the implementation is appropriate for engine integration and follows our requirements and guidelines, it can be integrated if and when there's reason to
